vmm/modules/vtd/vtd.c

```diff
@@ -363,9 +393,14 @@ void vtd_init(void)
 	vtd_init_dev_mapping();
 
 	event_register(EVENT_GPM_SET, vtd_update_domain_mapping);
-	event_register(EVENT_GUEST_MODULE_INIT, vtd_guest_setup);
 	event_register(EVENT_RESUME_FROM_S3, vtd_reactivate_from_s3);
 	event_register(EVENT_GPM_INVALIDATE, vtd_invalidate_cache);
+#ifndef ACTIVATE_VTD_BY_VMCALL
+	event_register(EVENT_GUEST_MODULE_INIT, vtd_guest_setup);
+#endif
+
+#define VMCALL_ACTIVATE_VTD 0x56544400ULL        // "VTD"
+	vmcall_register(GUEST_REE, VMCALL_ACTIVATE_VTD, vmcall_activate_vtd);
```

Comment:
register vmcall only when ACTIVATE_VTD_BY_VMCALL is set.
BTW, in this case, we can move the global variable "vtd_activated" into vmcall_activate_vtd(), as static local variable.